src/libopenrave/environment.cpp

```diff
@@ -183,124 +244,87 @@ void EnvironmentBase::EnvironmentBaseInfo::DeserializeJSONWithMapping(const rapi
-                pRobotBaseInfo->DeserializeJSON(rKinBodyInfo, fUnitScale, options);
-                pRobotBaseInfo->_id = id;
+                _DeserializeKinBodyInfo<RobotBase::RobotBaseInfo>(
+                    rKinBodyInfo, fUnitScale, options, isDeleted, existingBodyIndex, pId,
+                    /* mutable */ _vBodyInfos, /* mutable */ existingBodyInfoIndicesById, /* mutable */ existingBodyInfoIndicesByName);
             }
             else {
-                // not a robot
-                if (itExistingBodyInfo == _vBodyInfos.end()) {
-                    // in case no such id
-                    if (!isDeleted) {
-                        KinBody::KinBodyInfoPtr pKinBodyInfo(new KinBody::KinBodyInfo());
-                        pKinBodyInfo->DeserializeJSON(rKinBodyInfo, fUnitScale, options);
-                        if (!pKinBodyInfo->_name.empty()) {
-                            pKinBodyInfo->_id = id;
-                            _vBodyInfos.push_back(pKinBodyInfo);
-                            RAVELOG_VERBOSE_FORMAT("created new body id='%s'", id);
-                        } else {
-                            RAVELOG_WARN_FORMAT("new body id='%s' does not have a name, so skip creating", id);
-                        }
-                    }
-                    continue;
-                }
-                // in case same id exists before
-                if (isDeleted) {
-                    RAVELOG_VERBOSE_FORMAT("deleted body id='%s'", id);
-                    _vBodyInfos.erase(itExistingBodyInfo);
-                    continue;
-                }
-                KinBody::KinBodyInfoPtr pKinBodyInfo = *itExistingBodyInfo;
-                RobotBase::RobotBaseInfoPtr pRobotBaseInfo = OPENRAVE_DYNAMIC_POINTER_CAST<RobotBase::RobotBaseInfo>(pKinBodyInfo);
-                if (!!pRobotBaseInfo) {
-                    // previous body was a robot
-                    // need to replace with a new KinBodyInfo
-                    pKinBodyInfo.reset(new KinBody::KinBodyInfo());
-                    *itExistingBodyInfo = pKinBodyInfo;
-                    *pKinBodyInfo = *((KinBody::KinBodyInfo*)pRobotBaseInfo.get());
-                    RAVELOG_VERBOSE_FORMAT("replaced robot as a body id='%s'", id);
-                }
-                pKinBodyInfo->DeserializeJSON(rKinBodyInfo, fUnitScale, options);
-                pKinBodyInfo->_id = id;
+                _DeserializeKinBodyInfo<KinBody::KinBodyInfo>(
+                    rKinBodyInfo, fUnitScale, options, isDeleted, existingBodyIndex, pId,
+                    /* mutable */ _vBodyInfos, /* mutable */ existingBodyInfoIndicesById, /* mutable */ existingBodyInfoIndicesByName);
             }
         }
+
+        // Now that we're done processing, filter _vBodyInfos to remove any holes left from deleting bodies
+        _vBodyInfos.erase(std::remove_if(_vBodyInfos.begin(), _vBodyInfos.end(), [](const KinBody::KinBodyInfoPtr& p) { return !p; }), _vBodyInfos.end());
```

Comment:
Can you re-write this in an optimized way without using lambda functions and std::remove_if?

Reply:
Is the objection `remove_if` or the lambda? `remove_if` is already O(N), there isn't much reason to re-write it. Would you accept a struct based predicate instead? I have to say that the aversion to lambdas remains difficult to understand. 
```diff
         // Now that we're done processing, filter _vBodyInfos to remove any holes left from deleting bodies
-        _vBodyInfos.erase(std::remove_if(_vBodyInfos.begin(), _vBodyInfos.end(), [](const KinBody::KinBodyInfoPtr& p) { return !p; }), _vBodyInfos.end());
+        struct IsValidPointer {
+            bool operator()(const KinBody::KinBodyInfoPtr& p) { return !!p; }
+        };
+        _vBodyInfos.erase(std::remove_if(_vBodyInfos.begin(), _vBodyInfos.end(), IsValidPointer{}), _vBodyInfos.end());
```

Reply:
Because usually there's a way to write clearer code without lambdas..
```c++
_vBodyInfos.erase(std::remove(_vBodyInfos.begin(), _vBodyInfos.end(), KinBody::KinBodyInfoPtr()), _vBodyInfos.end());
```